### PR TITLE
Improve flag rendering performance, also without flags

### PR DESCRIPTION
**Pull request by Robyt3**
Cache the current and previous `CNetObj_Flag` snap objects in the `CSnapState`, instead of iterating over all snap items in the current snapshot each time that items including flags are being rendered.

Also cache the previous `CNetObj_GameData` snap object in the `CSnapState` in addition to the current one, to avoid searching for this snap item multiple times. Thereby, the `m_GameDataSnapId` member is not necessary anymore.

On the map dm1 with 64 players at zoom 0, this decreases the average frametime from 1431 µs to 1180 µs. The CPU time spent in the `CItems::OnRender` function in this scenario is reduced from 9.14% to 1.02% of the total CPU time.

Frametime comparison:
- 
- 

Profiling comparison:
- Before: 
- After: 

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/game/client/gameclient.h [hunk 1]
```diff
@@ -329,9 +329,10 @@ class CGameClient : public IGameClient
 		const CNetObj_SpectatorInfo *m_pSpectatorInfo;
 		const CNetObj_SpectatorInfo *m_pPrevSpectatorInfo;
 		const CNetObj_Flag *m_apFlags[2];
+		const CNetObj_Flag *m_apPrevFlags[2];
```
Comment (ChillerDragon) [inline src/game/client/gameclient.h]: While at it we could go with NUM_TEAMS here instead of 2

Comment (ChillerDragon) [reply src/game/client/gameclient.h]: Ah never mind ddnet broke that. Only `protocol7::NUM_TEAMS` would work and thats weird. Ignore what I said.

**Comment by ChillerDragon** [comment]
I had a quick look at my fps. And can not see any change. So no regression yay. Also no improvement :thinking: 

I closed everything. Had only the client running nothing else. Not even the server. The server was on a different machine in the local network.

<details>

screenshot_2025-09-09_13-27-31
screenshot_2025-09-09_13-28-32
screenshot_2025-09-09_13-29-40
screenshot_2025-09-09_13-30-40
screenshot_2025-09-09_13-34-23
screenshot_2025-09-09_13-35-01
screenshot_2025-09-09_13-34-29
screenshot_2025-09-09_13-35-06

</details>

The fps shown in the screenshot don't say much because the fps were jumping up and down by like 100~ but i saw no tendency up or downwards when comparing with my bare eyes.

---

Sadly it also breaks my nice multi flag mod. So with no visible fps improvement and multi flag rendering regression I don't think this pr is a overall improvement.

image

**Comment by Learath2** [comment]
Maybe your GPU cant keep up. Or maybe your settings are messed up. I don't see why this wouldn't lead to an improvement on some systems

**Comment by ChillerDragon** [comment]
My GPU is a NVIDIA GeForce RTX 4090 I tested this in 4k and use gnome with X11 and the non free nvidia drivers (nvidia 575.57.08-2). As I said as long as there is no regression I am fine with it. If it improves on some system even a bit thats a win. But I would not trade functionality for it. If this removes the option to render multiple flags and limits server mods then it has to at least tripple fps on all systems.

**Comment by Robyt3** [comment]
> Sadly it also breaks my nice multi flag mod. So with no visible fps improvement and multi flag rendering regression I don't think this pr is a overall improvement.

I almost expected that mods would somehow break from the check, but I didn't consider they would just try to show multiple flags. We could remove the snap ID validation again and instead store a vector of all flags. Then mods could use as many flags as before but it would still avoid the linear search when rendering.

> Maybe your GPU cant keep up. Or maybe your settings are messed up. I don't see why this wouldn't lead to an improvement on some systems

With as many FPS as you have in the screenshots, I also didn't see a large change. The improvement might be larger if you are more limited by CPU.

**Comment by Robyt3** [comment]
Can't use `std::vector` inside `CSnapState` because we `mem_zero` it, so I'll probably have to refactor that first.

**Comment by Robyt3** [comment]
Updated to use an array to support up to 1024 flags and removed the flag snap object ID validation, so mods can continue to use multiple flags. Performance without flags seems unchanged compared to previous profiling with a maximum of 2 flags.

**Comment by ChillerDragon** [comment]
I gave it a quick test on a way weaker laptop than the last test device. The flag rendering behavior seems unchanged :rocket: thanks for fixing that :heart: 

## master_8274a845bcb793

master_8274a845bcb793

## rob_itemop_7211e9961c60

rob_itemop_7211e9961c60

As you can see in the top right fps counter this time I could measure an fps improvement. And this is also not a minor hiccup. I basically have steady 200 fps on master and steady over 1000 fps with this pr. On a map with many flags and a low end device.

Thats some quality noticeable fps improvement :rocket: 

The fact that it changed from `const CNetObj_Flag *m_apFlags[2];` to `const CNetObj_Flag *m_apFlags[1024];` is a bit weird because there are only two teams and only two flags. I do not know any mod that actually makes use of displaying multiple flags. But somehow it was possible. I think even if weird the code now communicates cleaner that the client is capable of rendering so many flags. Also the seemingly arbitrary limit of 1024 should to be close enough to the current maximum snap size so that nobody will run into any limits here. This is something I asked for on discord because I was concern about a `std::vector` doing performance affecting allocations in some edge case we forget to consider. Because we have a upper limit given by the max snap size using a fixed size array seems okay.

All in all I did not test the latest version very thoroughly, but so far I assume this pr is good. Improving fps without any draw backs. I will give some other maintainer time to have a look. Because this is the kind of change that is better not rushed.

**Comment by ChillerDragon** [comment]
Can change the pr title from

**Improve flag rendering performance, especially without flags**

to

**Improve flag rendering performance, especially with many flags**

lmao

File: src/game/client/gameclient.h [hunk 2]
```diff
@@ -329,10 +329,12 @@ class CGameClient : public IGameClient
 		const CNetObj_PlayerInfo *m_pLocalInfo;
 		const CNetObj_SpectatorInfo *m_pSpectatorInfo;
 		const CNetObj_SpectatorInfo *m_pPrevSpectatorInfo;
-		const CNetObj_Flag *m_apFlags[2];
+		int m_NumFlags;
+		const CNetObj_Flag *m_apFlags[1024];
+		const CNetObj_Flag *m_apPrevFlags[1024];
```
Comment (KebsCS) [inline src/game/client/gameclient.h]:
```suggestion
		const CNetObj_Flag *m_apFlags[CSnapshot::MAX_ITEMS];
		const CNetObj_Flag *m_apPrevFlags[CSnapshot::MAX_ITEMS];
```
Or where does the 1024 come from?

Comment (ChillerDragon) [reply src/game/client/gameclient.h]: `CSnapshot::MAX_ITEMS` is 1024 oh thats nice. Yes we can use that. I tested adding flags and my server capped at 1010 with one pickup and tee rendered so thats how we came up with 1024

Comment (Robyt3) [reply src/game/client/gameclient.h]: Good call. I picked 1024 arbitrarily as the next round number, but `CSnapshot::MAX_ITEMS` captures the intent more clearly.

**Comment by ChillerDragon** [comment]
Tested on another laptop which was at 200-400 fps depending on map and zoom and its fps were also not affected at all by this pr. But the fps TRIPPLE I had on the weak laptop is huuge!!! And I guess you also measured some improvements @Robyt3. Which means so far no fps decrease was measured only increases :rocket: :rocket: :rocket:

Looks like big performance offenders were also the snap lookups. Makes me wanna push #8037 again.

Thanks @Robyt3 this is good stuff!
